stores/metrics.go

```diff
@@ -517,36 +517,49 @@ func normaliseTimestamp(start time.Time, interval time.Duration, t unixTimeMS) u
 	return unixTimeMS(time.UnixMilli(normalizedMS))
 }
 
-func roundPeriodExpr(db *gorm.DB, start time.Time, interval time.Duration) clause.Expr {
-	if !isSQLite(db) {
-		return gorm.Expr("CAST(FLOOR((timestamp - ?) / ?) * ? AS SIGNED)", unixTimeMS(start), interval.Milliseconds(), interval.Milliseconds())
-	} else {
-		return gorm.Expr("(timestamp - ?) / ? * ?", unixTimeMS(start), interval.Milliseconds(), interval.Milliseconds())
-	}
-}
-
 func (s *SQLStore) findAggregatedContractPeriods(start time.Time, n uint64, interval time.Duration) ([]dbContractMetric, error) {
+	if n > api.MaxMetricIntervals {
+		return nil, api.ErrMaxIntervalsExceeded
+	}
 	end := start.Add(time.Duration(n) * interval)
 	var metricsWithPeriod []struct {
 		Metric dbContractMetric `gorm:"embedded"`
 		Period int64
 	}
-	currentPeriod := int64(math.MinInt64)
 	err := s.dbMetrics.Raw(`
-		SELECT * FROM contracts
+		WITH RECURSIVE periods AS (
+			SELECT ? AS period_start
+			UNION ALL
+			SELECT period_start + ?
+			FROM periods
+			WHERE period_start < ? - ?
+		)
+		SELECT contracts.*, i.Period FROM contracts
 		INNER JOIN (
-			SELECT fcid, MIN(timestamp) as timestamp, ? AS Period
-			FROM contracts
-			WHERE timestamp >= ? AND timestamp < ?
-			GROUP BY Period, fcid) i
-		ON contracts.fcid = i.fcid AND contracts.timestamp = i.timestamp
-	`, roundPeriodExpr(s.dbMetrics, start, interval),
-		unixTimeMS(start), unixTimeMS(end)).
+		SELECT
+			p.period_start as Period, 
+			MIN(c.id) AS id
+		FROM
+			periods p
+		INNER JOIN
+			contracts c ON c.timestamp >= p.period_start AND c.timestamp < p.period_start + ?
+		GROUP BY
+			p.period_start, c.fcid
+		ORDER BY
+			p.period_start ASC
+		) i ON contracts.id = i.id
+	`, unixTimeMS(start),
```

Comment:
Is the period start rounded down to the nearest interval? Seems you'd get different periods if you compare to what we used to do in `roundPeriodExpr`

Reply:
The periods don't in the query don't really matter as long as they are distinct and contracts are associated correctly with them. `normaliseTimestamp` later updates the timestamp of each returned metric to be correct.